evmcore/validation.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2014 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package evmcore
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/0xsoniclabs/sonic/gossip/gasprice/gaspricelimits"
+	"github.com/0xsoniclabs/sonic/utils"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+// validationOptions is a set of options to adjust the validation of transactions
+// according to the current state of the transaction pool.
+type validationOptions struct {
+	istanbul bool // Fork indicator whether we are in the istanbul revision.
+	shanghai bool // Fork indicator whether we are in the shanghai revision.
+
+	eip2718 bool // Fork indicator whether we are using EIP-2718 type transactions.
+	eip1559 bool // Fork indicator whether we are using EIP-1559 type transactions.
+	eip4844 bool // Fork indicator whether we are using EIP-4844 type transactions.
+	eip7623 bool // Fork indicator whether we are using EIP-7623 floor gas validation.
+	eip7702 bool // Fork indicator whether we are using EIP-7702 set code transactions.
+
+	currentState   TxPoolStateDB // Current state in the blockchain head
+	currentMaxGas  uint64        // Current gas limit for transaction caps
+	currentBaseFee *big.Int      // Current base fee for transaction caps
+	minTip         *big.Int      // Minimum gas tip to enforce for acceptance into the pool
+
+	locals  *accountSet // Set of local transaction to exempt from eviction rules
+	isLocal bool        // Whether the transaction came from a local source
+
+}
+
+// validateTx checks whether a transaction is valid according to the current
+// options and adheres to some heuristic limits of the local node (price and size).
+func validateTx(tx *types.Transaction, signer types.Signer, opt validationOptions) error {
```

Comment:
nit: include the `signer` in the options?

Reply:
https://github.com/0xsoniclabs/sonic/pull/154/commits/9ea293d2a5d32b4d447845935fa88a7397446828